### ISSUE TITLE
Type mismatch with type alias and associated types

### ISSUE THREAD

**Issue by arcnmx** [rust-lang/rust#34738]
Type aliases using associated types / `as` to resolve to a concrete type do not work as expected:

``` rust
trait Tr {
    type Ty;
    fn f(self) -> Self::Ty;
}

impl Tr for () {
    type Ty = f32;

    fn f(self) -> Self::Ty { 0.5 }
}

type TrTy<T> = <T as Tr>::Ty;

fn main() {
    let _: <() as Tr>::Ty = Tr::f(()); // works
    let _: TrTy<()> = Tr::f(()); // error
}
```

```
error: mismatched types [--explain E0308]
  --> <anon>:16:23
16 |>     let _: TrTy<()> = Tr::f(()); // error
   |>                       ^^^^^^^^^ expected associated type, found f32
note: expected type `<() as Tr>::Ty`
note:    found type `f32`
```

[playpen repro](https://is.gd/oECsXf)

**Comment by TimNN** [comment]
At the very least related to, but probably a dupe of #28828.

**Comment by sorpaas** [comment]
Any workaround for this issue?

**Comment by Mark-Simulacrum** [comment]
Example given compiles today. Tests were added in fix for #28828 which seem to cover this, closing.
